internal/controllers/topology/cluster/patches/template.go

```diff
@@ -18,123 +18,115 @@ package patches
+	// Find the requestItem.
+	requestItem := getRequestItem(req, holderKind, holderFieldPath, mdTopologyName)
 
-	// If a patch doesn't apply to any object, this is a misconfiguration.
-	if template == nil {
-		return nil, errors.Errorf("failed to get template %s", templateRef)
+	if requestItem == nil {
+		return nil, errors.Errorf("failed to get request item with holder kind %q, holder field path %q and MD topology name %q", holderKind, holderFieldPath, mdTopologyName)
+	}
+
+	// Unmarshal the template.
+	template, err := bytesToUnstructured(requestItem.Object.Raw)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to convert template to Unstructured")
 	}
 
-	return templateToUnstructured(template)
+	return template, nil
 }
 
-// getTemplate is a utility func that returns a template matching the templateRef from a GenerateRequest.
-func getTemplate(req *api.GenerateRequest, templateRef api.TemplateRef) *api.GenerateRequestTemplate {
-	for _, template := range req.Items {
-		if templateRefsAreEqual(templateRef, template.TemplateRef) {
-			return template
+// getRequestItemByUID is a utility func that returns a template matching the uid from a GeneratePatchesRequest.
+func getRequestItemByUID(req *runtimehooksv1.GeneratePatchesRequest, uid types.UID) *runtimehooksv1.GeneratePatchesRequestItem {
+	for i := range req.Items {
+		if req.Items[i].UID == uid {
+			return &req.Items[i]
 		}
 	}
 	return nil
 }
 
-// templateRefsAreEqual returns true if the TemplateRefs are equal.
-func templateRefsAreEqual(a, b api.TemplateRef) bool {
-	return a.APIVersion == b.APIVersion &&
-		a.Kind == b.Kind &&
-		a.TemplateType == b.TemplateType &&
-		a.MachineDeploymentRef.TopologyName == b.MachineDeploymentRef.TopologyName &&
-		a.MachineDeploymentRef.Class == b.MachineDeploymentRef.Class
-}
+// getRequestItem is a utility func that returns a template matching the holderKind, holderFiledPath and mdTopologyName from a GeneratePatchesRequest.
+func getRequestItem(req *runtimehooksv1.GeneratePatchesRequest, holderKind, holderFieldPath, mdTopologyName string) *runtimehooksv1.GeneratePatchesRequestItem {
+	for _, template := range req.Items {
+		if holderKind != "" && template.HolderReference.Kind != holderKind {
+			continue
+		}
+		if holderFieldPath != "" && template.HolderReference.FieldPath != holderFieldPath {
+			continue
+		}
+		if mdTopologyName != "" {
+			templateVariables := toMap(template.Variables)
 
-// templateToUnstructured converts a GenerateRequestTemplate into an Unstructured object.
-func templateToUnstructured(t *api.GenerateRequestTemplate) (*unstructured.Unstructured, error) {
-	// Unmarshal the template.
-	u, err := bytesToUnstructured(t.Template.Raw)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to convert template to Unstructured")
-	}
+			v, err := variables.GetVariableValue(templateVariables, "builtin.machineDeployment.topologyName")
```

Comment:
nit: we should introduce constants for builtin variables we use across the codebase
(could be a follow up)

Reply:
I'll make it a follow-up (mostly because we have a full list of builtin variables in patch_validation and I'll separate out the discussion how to ideally fix this to a non-critical-path PR.

Reply:
Added below cleanups under: https://github.com/kubernetes-sigs/cluster-api/issues/6330
(just to collect for now, I'll split up as separate issue later if appropriate)